processing/src/main/java/io/druid/segment/IndexMerger.java

```diff
@@ -305,125 +300,75 @@ public int seek(int dictId)
     }
   }
 
-  class MMappedIndexRowIterable implements Iterable<Rowboat>
+  static TransformableRowIterator toMergedIndexRowIterator(
+      TransformableRowIterator sourceRowIterator,
+      int indexNumber,
+      final List<DimensionMerger> mergers
+  )
   {
-    private final Iterable<Rowboat> index;
-    private final List<String> convertedDims;
-    private final int indexNumber;
-    private final List<DimensionMerger> mergers;
-
-
-    MMappedIndexRowIterable(
-        Iterable<Rowboat> index,
-        List<String> convertedDims,
-        int indexNumber,
-        final List<DimensionMerger> mergers
-    )
-    {
-      this.index = index;
-      this.convertedDims = convertedDims;
-      this.indexNumber = indexNumber;
-      this.mergers = mergers;
-    }
-
-    @Override
-    public Iterator<Rowboat> iterator()
-    {
-      return Iterators.transform(
-          index.iterator(),
-          new Function<Rowboat, Rowboat>()
-          {
-            @Override
-            public Rowboat apply(@Nullable Rowboat input)
-            {
-              Object[] dims = input.getDims();
-              Object[] newDims = new Object[convertedDims.size()];
-              for (int i = 0; i < convertedDims.size(); ++i) {
-                if (i >= dims.length) {
-                  continue;
-                }
-                newDims[i] = mergers.get(i).convertSegmentRowValuesToMergedRowValues(dims[i], indexNumber);
-              }
-
-              final Rowboat retVal = new Rowboat(
-                  input.getTimestamp(),
-                  newDims,
-                  input.getMetrics(),
-                  input.getRowNum(),
-                  input.getHandlers()
-              );
-
-              retVal.addRow(indexNumber, input.getRowNum());
-
-              return retVal;
-            }
-          }
+    RowPointer sourceRowPointer = sourceRowIterator.getPointer();
+    boolean anySelectorChanged = false;
+    ColumnValueSelector[] convertedDimensionSelectors = new ColumnValueSelector[mergers.size()];
+    ColumnValueSelector[] convertedMarkedDimensionSelectors = new ColumnValueSelector[mergers.size()];
+    for (int i = 0; i < mergers.size(); i++) {
+      ColumnValueSelector sourceDimensionSelector = sourceRowPointer.getDimensionSelector(i);
+      ColumnValueSelector convertedDimensionSelector =
+          mergers.get(i).convertSegmentRowValuesToMergedRowValues(indexNumber, sourceDimensionSelector);
+      convertedDimensionSelectors[i] = convertedDimensionSelector;
+      //noinspection ObjectEquality
+      anySelectorChanged |= convertedDimensionSelector != sourceDimensionSelector;
+
+      convertedMarkedDimensionSelectors[i] = mergers.get(i).convertSegmentRowValuesToMergedRowValues(
+          indexNumber,
+          sourceRowIterator.getMarkedPointer().dimensionSelectors[i]
```

Comment:
please add some comments to this method to make understanding the flow easier.

Reply:
Fixed. However, testing this is not the subject of this PR. Please add tests that ensure that columns are merged correctly in the presence of nulls in one of your null-handling PRs.

Reply:
Reorganized code a bit and added comments.